### PR TITLE
Add Base64 support for FileStorage

### PR DESCRIPTION
**Pull request by wiryls**
[GSoC 2016] FileStorage.

### What does this PR change?

- Add Base64 support for reading and writing XML\YML file.
```cpp
// The two new functions for writing data:
void cvWriteRawData_Base64(cv::FileStorage & fs, const void* _data, int
len, const char* dt);
void cvWriteMat_Base64(cv::FileStorage & fs, cv::String const & name,
cv::Mat const & mat);
```

- Change YML file header form `YAML:1.0` to `YAML 1.0`. (standard format)
- Add test for Base64 part.

### CODE REVIEW

**Comment by vpisarev** [comment]
thanks! I will check it. Quick question - you've changed the header in YAML files written by OpenCV. Does the decoder support both the new and the old format? We want to be able to read all the previously stored .yml files.

**Comment by wiryls** [comment]
Yes, it supports both the new and the old format.
I plan to add a test file for yml compatibility testing.

File: modules/core/include/opencv2/core/persistence.hpp [hunk 1]
```diff
@@ -1238,6 +1238,11 @@ inline String::String(const FileNode& fn): cstr_(0), len_(0) { read(fn, *this, *
 
 //! @endcond
 
+
+CV_EXPORTS void cvWriteRawData_Base64(::cv::FileStorage & fs, const void* _data, int len, const char* dt);
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/persistence.hpp]: please, change the API the first parameter to "CvFileStorage* fs"

File: modules/core/include/opencv2/core/persistence.hpp [hunk 2]
```diff
@@ -1238,6 +1238,11 @@ inline String::String(const FileNode& fn): cstr_(0), len_(0) { read(fn, *this, *
...
+
+CV_EXPORTS void cvWriteMat_Base64(::cv::FileStorage & fs, ::cv::String const & name, ::cv::Mat const & mat);
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/persistence.hpp]: please, change the API to

<pre>
CV_EXPORTS void cvWriteMat_Base64(CvFileStorage* fs, const char* name, const CvMat* mat);
</pre>

File: modules/core/include/opencv2/core/persistence.hpp (hunk 1 above)
Comment (wiryls) [inline modules/core/include/opencv2/core/persistence.hpp]: Hello! About changing this API, there is one problem to be discussed.

`cvWriteRawData_Base64` cannot be continually called, like:
```
// in `icvWriteMat`
for( y = 0; y < size.height; y++ )
    cvWriteRawData( fs, mat->data.ptr + (size_t)y*mat->step, size.width, dt );
```

When doing something like this, the output data cannot be merged into a single one. To solve this, I might need to add a private variable to `CvFileStorage` for recording status. Meanwhile, users must call `cvStartWriteStruct(fs, name, CV_NODE_SEQ, "binary")` themselves. It doesn't look good.

Another way, I can change the parameters to `(CvFileStorage * fs, const char* name, const void* _data, int len, const char* dt)` and treat it as a complete data type. But then this function must output all the data at once. Is this ok?

**Comment by vpisarev** [comment]
@wiryls, probably, we should have 3 functions then: cvStartWriteRawData_Base64(), cvWriteRawData_Base64() and cvEndWriteRawData_Base64()

**Comment by wiryls** [comment]
@vpisarev  Ok, and can I add a member variable to `CvFileStorage`?
 
Because I need something to hold the length of binary data and data whose length is not a multiple of 3 (the part that temporarily should not be encoded). Or is there any better way to achieve it?

**Comment by vpisarev** [comment]
@wiryls, yes, sure, add all the necessary data to CvFileStorage

**Comment by wiryls** [comment]
@vpisarev . I've done that.

`cvWriteRawData_Base64` now become three functions:

```cpp
void cvStartWriteRawData_Base64(::CvFileStorage * fs, const char* name, int len, const char* dt);
void cvWriteRawData_Base64(::CvFileStorage * fs, const void* _data, int len);
void cvEndWriteRawData_Base64(::CvFileStorage * fs);
```

By the way, I've also encountered a strange error when building on OpenCL (maybe it's a bug). See [commit 677d4d20ce9de6b548887a05bbdd0ffb207281b0](http://pullrequest.opencv.org/buildbot/builders/precommit_opencl/builds/5721). And build success after I retry.

**Comment by vpisarev** [comment]
:+1:

**Pull request merged**